sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -477,6 +478,25 @@ object RemoveRedundantAliases extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = removeRedundantAliases(plan, AttributeSet.empty)
 }
 
+/**
+ * Remove redundant aggregates from a query plan. A redundant aggregate is an aggregate whose
+ * only goal is to keep distinct values, while its parent aggregate would ignore duplicate values.
+ */
+object RemoveRedundantAggregates extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
+    case upper @ Aggregate(_, _, lower: Aggregate) if lowerIsRedundant(upper, lower) =>
+      upper.copy(child = lower.child)
+  }
+
+  private def lowerIsRedundant(upper: Aggregate, lower: Aggregate): Boolean = {
+    val upperReferencesOnlyGrouping = upper.references
+      .subsetOf(AttributeSet(lower.groupingExpressions))
```

Comment:
I do not have a strong understanding of aliases.

Is there a way this can be improved to also handle aliases:
```
    val query = relation
      .groupBy(('a + 'b) as 'c)(('a + 'b) as 'c, count('b))
      .groupBy('c)('c)
      .analyze
    val expected = relation
      .groupBy(('a + 'b) as 'c)(('a + 'b) as 'c)
      .analyze
```

Perhaps @maropu or @cloud-fan could advice me here

Reply:
I think `CleanupAliases` removes aliases in `groupingExpressions` in the analyzer phase.

```
    val upperReferencesOnlyGrouping = upper.references
      .subsetOf(AttributeSet(lower.groupingExpressions))
```
How about checking if upper references are a subset of non-aggregate expressions in a lower node instead?

Reply:
Done